staging/src/k8s.io/component-base/logs/example/slog2k8s/slog2k8s.go

```diff
@@ -0,0 +1,51 @@
+//go:build go1.21
+// +build go1.21
+
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// slog2k8s demonstrates how an application using log/slog for logging
+// can include Kubernetes packages.
+package main
+
+import (
+	"context"
+	"log/slog"
+	"os"
+
+	"github.com/go-logr/logr/slogr"
+	"k8s.io/klog/v2"
+)
+
+func main() {
+	options := slog.HandlerOptions{AddSource: true}
+	textHandler := slog.NewTextHandler(os.Stderr, &options)
+	textLogger := slog.New(textHandler)
+
+	// Use text output as default logger.
+	slog.SetDefault(textLogger)
+
+	// This also needs to be done through klog to ensure that all code
+	// using klog uses the text handler. klog.Background/TODO/FromContext
+	// will return a thin wrapper around the textHandler, so all that klog
+	// still does is manage the global default and retrieval from contexts.
+	klog.SetLoggerWithOptions(slogr.NewLogr(textHandler), klog.ContextualLogger(true))
```

Comment:
It's not nice that klog-based packages only use slog when explicitly configured to do so. It adds a direct dependency on klog to applications which otherwise don't use it.

The alternative is to make logr and klog aware of slog:
- `klog.Background` could use `slog.Default` and convert to a `log.Logger` with `slogr.NewLogr`. Note that this would imply that `klog.Info` *also* would have to use slog by default, otherwise `klog.Info` produces different output than `klog.Background().Info()` - this would be a pretty big breaking change.
- `klog.FromContext` could be replaced in e.g. `client-go` with `logr.FromContext`, and that could support both `logr.Logger` and `slog.Handler` by converting as needed. `slog.Default` again would become the fallback. Again, this is a pretty big change in behavior and implies adding new API calls to logr for `SlogHandlerFromContext`.

I don't think that we should do this. I'm just pointing out the alternative to show that it isn't easy.

Reply:
> Why not offer klog interfaces to point kubernetes logging at either logr

We already have that, with `klog.SetLogger`.

> or slog

That could be useful. How about this:
```
// SetSlogLogger reconfigures klog to log through the slog logger. The logger must not be nil.
func SetSlogLogger(logger *slog.Logger) {
   SetLoggerWithOptions(slogr.NewLogr(logger.Handler()), klog.ContextualLogger(true))
}
```

I briefly considered allowing nil and then using `slog.Default()` as fallback, but that leads to a gotcha:
```
klog.SetSlogLogger(nil)
slog.SetDefault(<some new logger>)
```
would not make klog use that new logger.

Reply:
And this shouldn't be using `slogr` :grimacing: ...

Reply:
klog now has `SetSlogLogger` which is now used in this example.